File: apps/mobile-android/app/src/main/java/com/univalle/pedrochacolla/ui/dashboard/MapCloudShadowEffect.kt
package com.univalle.pedrochacolla.ui.dashboard

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RadialGradient
import android.graphics.Shader
import com.univalle.pedrochacolla.utils.map.AmbientTickOptions
import com.univalle.pedrochacolla.utils.map.MapAmbientZone
import kotlin.math.cos
import kotlin.math.sin
import kotlin.random.Random

class MapCloudShadowEffect {
    var intensity = 0.4f
    var sizeMul = 1f
        set(value) { field = value.coerceIn(0.08f, 2.5f) }

    private data class Blob(
        var bx: Float, var by: Float, val r: Float, val layer: Int, var phase: Float,
    )

    private val blobs = mutableListOf<Blob>()
    private var containsPoint: ((Float, Float) -> Boolean)? = null
    private var driftPhase = 0f
    private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    fun clear() { blobs.clear(); driftPhase = 0f }
    fun setContainsPoint(fn: ((Float, Float) -> Boolean)?) { containsPoint = fn }

    private fun inZone(bx: Float, by: Float) = containsPoint?.invoke(bx, by) ?: true

    private fun spawn(bounds: AmbientTickOptions, layer: Int): Blob {
        val b = bounds.bounds
        val spanX = b.maxX - b.minX
        val spanY = b.maxY - b.minY
        var bx = b.minX + Random.nextFloat() * spanX
        var by = b.minY + Random.nextFloat() * spanY
        repeat(16) {
            bx = b.minX + Random.nextFloat() * spanX
            by = b.minY + Random.nextFloat() * spanY
            if (inZone(bx, by)) return@repeat
        }
        val baseR = if (layer == 0) MapAmbientZone.parkPlanSize(32f + Random.nextFloat() * 48f)
        else MapAmbientZone.parkPlanSize(16f + Random.nextFloat() * 26f)
        return Blob(bx, by, baseR, layer, Random.nextFloat() * 6.28f)
    }

    fun tick(options: AmbientTickOptions, dt: Float = 1f) {
        options.containsPoint?.let { containsPoint = it }
        val b = options.bounds
        if (b.maxX - b.minX <= 0f || b.maxY - b.minY <= 0f) return
        val inten = maxOf(0.15f, intensity)
        val target = (2 + inten * 4).toInt() + (2 + inten * 5).toInt()
        while (blobs.size < target) blobs.add(spawn(options, blobs.size % 2))
        while (blobs.size > target) blobs.removeAt(blobs.lastIndex)
        driftPhase += 0.004f * dt
        val (wx, wy) = MapAmbientZone.ambientWindVector(options.wind)
        val windX = wx * 0.35f + sin(driftPhase) * 0.08f
        val windY = wy * 0.28f + cos(driftPhase * 0.7f) * 0.05f
        for (blob in blobs) {
            val parallax = if (blob.layer == 0) 0.35f else 0.85f
            blob.phase += 0.01f * dt
            blob.bx += (windX + sin(blob.phase) * 0.04f) * parallax * dt
            blob.by += (windY + cos(blob.phase * 0.9f) * 0.03f) * parallax * dt
            if (!inZone(blob.bx, blob.by) || blob.bx < b.minX - 40f || blob.bx > b.maxX + 40f) {
                val n = spawn(options, blob.layer)
                blob.bx = n.bx; blob.by = n.by; blob.phase = n.phase
            }
        }
    }

    fun draw(canvas: Canvas, clipPath: Path?, toScreen: (Float, Float) -> Pair<Float, Float>, screenScale: Float) {
        val inten = maxOf(0.2f, intensity)
        val sr = MapAmbientZone.ambientScreenScale(screenScale, sizeMul)
        canvas.save()
        clipPath?.let { canvas.clipPath(it) }
        for (b in blobs) {
            if (!inZone(b.bx, b.by)) continue
            val (x, y) = toScreen(b.bx, b.by)
            val pulse = 0.9f + sin(b.phase) * 0.1f
            val radius = b.r * sr * pulse
            val alpha = (inten * if (b.layer == 0) 0.14f else 0.22f * 255).toInt()
            fillPaint.shader = RadialGradient(
                x, y, radius,
                intArrayOf(
                    android.graphics.Color.argb((alpha * 0.9f).toInt(), 25, 35, 55),
                    android.graphics.Color.TRANSPARENT,
                ),
                floatArrayOf(0.1f, 1f),
                Shader.TileMode.CLAMP,
            )
            canvas.drawOval(x - radius, y - radius * 0.68f, x + radius, y + radius * 0.68f, fillPaint)
        }
        fillPaint.shader = null
        canvas.restore()
    }
}

File: apps/web-admin/src/app/features/map/utils/map-cloud-shadow-effect.ts
import type { AmbientTickOptions } from './map-ambient-zone';
import { parkPlanSize, parkParticleTarget } from './map-park-visual-scale';
import { ambientScreenScale } from './map-ambient-zone';
import { ambientWindVector, DEFAULT_AMBIENT_WIND } from './map-ambient-wind';

interface ShadowBlob {
  bx: number;
  by: number;
  r: number;
  layer: 0 | 1;
  phase: number;
}

/** Sombras de nubes lentas con parallax (capa lejana vs cercana). */
export class MapCloudShadowEffect {
  private blobs: ShadowBlob[] = [];
  private intensity = 0.4;
  private sizeMul = 1;
  private containsPoint: ((bx: number, by: number) => boolean) | null = null;
  private driftPhase = 0;

  setIntensity(value: number): void {
    this.intensity = Math.min(1, Math.max(0, value));
  }

  setSizeMul(value: number): void {
    this.sizeMul = Math.min(2.5, Math.max(0.08, value));
  }

  setContainsPoint(fn: ((bx: number, by: number) => boolean) | null): void {
    this.containsPoint = fn;
  }

  clear(): void {
    this.blobs = [];
    this.driftPhase = 0;
  }

  private inZone(bx: number, by: number): boolean {
    return !this.containsPoint || this.containsPoint(bx, by);
  }

  private spawn(bounds: AmbientTickOptions['bounds'], layer: 0 | 1): ShadowBlob {
    const spanX = bounds.maxX - bounds.minX;
    const spanY = bounds.maxY - bounds.minY;
    let bx = bounds.minX + Math.random() * spanX;
    let by = bounds.minY + Math.random() * spanY;
    for (let t = 0; t < 16; t++) {
      bx = bounds.minX + Math.random() * spanX;
      by = bounds.minY + Math.random() * spanY;
      if (this.inZone(bx, by)) break;
    }
    const baseR = layer === 0
      ? parkPlanSize(32 + Math.random() * 48)
      : parkPlanSize(16 + Math.random() * 26);
    return { bx, by, r: baseR, layer, phase: Math.random() * Math.PI * 2 };
  }

  tick(options: AmbientTickOptions, dt = 1): void {
    if (options.containsPoint) this.containsPoint = options.containsPoint;
    const bounds = options.bounds;
    const spanX = bounds.maxX - bounds.minX;
    const spanY = bounds.maxY - bounds.minY;
    if (spanX <= 0 || spanY <= 0) return;

    const inten = Math.max(0.15, this.intensity);
    const targetFar = Math.floor(2 + inten * 4);
    const targetNear = Math.floor(2 + inten * 5);
    const target = targetFar + targetNear;

    while (this.blobs.length < target) {
      const layer = (this.blobs.length % 2) as 0 | 1;
      this.blobs.push(this.spawn(bounds, layer));
    }
    while (this.blobs.length > target) this.blobs.pop();

    this.driftPhase += 0.004 * dt;
    const wv = ambientWindVector(options.wind ?? DEFAULT_AMBIENT_WIND);
    const windX = wv.vx * 0.35 + Math.sin(this.driftPhase) * 0.08;
    const windY = wv.vy * 0.28 + Math.cos(this.driftPhase * 0.7) * 0.05;

    for (const b of this.blobs) {
      const parallax = b.layer === 0 ? 0.35 : 0.85;
      b.phase += 0.01 * dt;
      b.bx += (windX + Math.sin(b.phase) * 0.04) * parallax * dt;
      b.by += (windY + Math.cos(b.phase * 0.9) * 0.03) * parallax * dt;

      if (!this.inZone(b.bx, b.by) || b.bx < bounds.minX - 40 || b.bx > bounds.maxX + 40) {
        const n = this.spawn(bounds, b.layer);
        Object.assign(b, n);
      }
    }
  }

  draw(
    ctx: CanvasRenderingContext2D,
    clipPath: Path2D | null,
    toScreen: (bx: number, by: number) => { x: number; y: number },
    screenScale = 1,
  ): void {
    const inten = Math.max(0.2, this.intensity);
    const sr = ambientScreenScale(screenScale, this.sizeMul);
    ctx.save();
    if (clipPath) ctx.clip(clipPath);

    for (const b of this.blobs) {
      if (!this.inZone(b.bx, b.by)) continue;
      const { x, y } = toScreen(b.bx, b.by);
      const pulse = 0.9 + Math.sin(b.phase) * 0.1;
      const radius = b.r * sr * pulse;
      const alpha = inten * (b.layer === 0 ? 0.14 : 0.22);
      const g = ctx.createRadialGradient(x, y, radius * 0.1, x, y, radius);
      g.addColorStop(0, `rgba(25, 35, 55, ${alpha * 0.9})`);
      g.addColorStop(0.55, `rgba(15, 25, 40, ${alpha * 0.55})`);
      g.addColorStop(1, 'rgba(10, 18, 30, 0)');
      ctx.fillStyle = g;
      ctx.beginPath();
      ctx.ellipse(x, y, radius, radius * 0.68, 0.15, 0, Math.PI * 2);
      ctx.fill();
    }

    ctx.restore();
  }
}
